### PR TITLE
Flex-wrap in CKStackLayoutComponent

### PR DESCRIPTION
**Pull request by antigp**
This is implement of flex-wrap in CKStackLayoutComponent
Work with Horizontal and Vertical Stack

unknown-1

### CODE REVIEW

**Comment by ocrickard** [comment]
Great! Really excited about this. To start off we're going to need comprehensive test coverage to make sure this matches the standard.

File: ComponentKit/LayoutComponents/CKStackPositionedLayout.mm [hunk 1]
```diff
@@ -43,11 +43,23 @@ static CKStackPositionedLayout stackedLayout(const CKStackLayoutComponentStyle &
   const auto largestChildCrossSize = it == unpositionedLayout.items.end() ? 0 : crossDimension(style.direction, it->layout.size);
   const auto minCrossSize = crossDimension(style.direction, constrainedSize.min);
   const auto maxCrossSize = crossDimension(style.direction, constrainedSize.max);
+  const auto maxStackSize = stackDimension(style.direction, constrainedSize.max);
+
   const CGFloat crossSize = MIN(MAX(minCrossSize, largestChildCrossSize), maxCrossSize);
 
+  CGFloat maxLineHieght = 0;
```
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: *maxLineHeight

File: ComponentKit/LayoutComponents/CKStackPositionedLayout.mm [hunk 2]
```diff
@@ -59,9 +71,12 @@ static CKStackPositionedLayout stackedLayout(const CKStackLayoutComponentStyle &
       l.layout,
     };
     p = p + directionPoint(style.direction, stackDimension(style.direction, l.layout.size) + l.child.spacingAfter, 0);
+    if(crossDimension(style.direction, l.layout.size) > maxLineHieght) {
```
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: This is computed even if flexWrap isn't enabled.

File: ComponentKit/LayoutComponents/CKStackPositionedLayout.mm (hunk 1 above)
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: Also, line height I don't think is a very good name. This algorithm is built to work in either horizontal or vertical directions, so I doubt this is a height in every case right? Height implies vertical directionality of the value.

File: ComponentKit/LayoutComponents/CKStackPositionedLayout.mm [hunk 3]
```diff
@@ -43,11 +43,23 @@ static CKStackPositionedLayout stackedLayout(const CKStackLayoutComponentStyle &
...
+      CGFloat stackSize = (style.direction == CKStackLayoutDirectionVertical) ? p.y : p.x;
+      if (stackSize + l.child.spacingBefore + style.spacing + stackDimension(style.direction, l.layout.size) + l.child.spacingAfter > maxStackSize) {
+        crossPosition = crossPosition + maxLineHieght;
+        maxLineHieght = 0;
+        p = directionPoint(style.direction, offset, crossPosition);
+      }
+    }
     p = p + directionPoint(style.direction, l.child.spacingBefore, 0);
     if (!first) {
       p = p + directionPoint(style.direction, style.spacing, 0);
```
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: What happens here with your algorithm? When we proceed to break the line are we still adding spacing before the element? Probably need a test for this.

File: ComponentKit/LayoutComponents/CKStackPositionedLayout.mm [hunk 4]
```diff
@@ -43,11 +43,23 @@ static CKStackPositionedLayout stackedLayout(const CKStackLayoutComponentStyle &
...
+  CGFloat maxLineHieght = 0;
+  CGFloat crossPosition = 0;
   CGPoint p = directionPoint(style.direction, offset, 0);
   BOOL first = YES;
   auto stackedChildren = CK::map(unpositionedLayout.items, [&](const CKStackUnpositionedItem &l) -> CKComponentLayoutChild {
+    if(!first && style.flexWrap == CKStackLayoutWrapFlexWrap) {
+      CGFloat stackSize = (style.direction == CKStackLayoutDirectionVertical) ? p.y : p.x;
+      if (stackSize + l.child.spacingBefore + style.spacing + stackDimension(style.direction, l.layout.size) + l.child.spacingAfter > maxStackSize) {
+        crossPosition = crossPosition + maxLineHieght;
+        maxLineHieght = 0;
```
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: You probably need a comment here explaining why you're setting to 0. It takes a bit of reading to realize you're resetting the max line height for the next line.

File: ComponentKit/LayoutComponents/CKStackPositionedLayout.mm [hunk 5]
```diff
@@ -59,9 +71,12 @@ static CKStackPositionedLayout stackedLayout(const CKStackLayoutComponentStyle &
       l.layout,
     };
     p = p + directionPoint(style.direction, stackDimension(style.direction, l.layout.size) + l.child.spacingAfter, 0);
+    if(crossDimension(style.direction, l.layout.size) > maxLineHieght) {
+        maxLineHieght = crossDimension(style.direction, l.layout.size);
+    }
     return c;
   });
-  return {stackedChildren, crossSize};
+  return {stackedChildren, style.flexWrap == CKStackLayoutWrapFlexWrap ? (maxLineHieght + crossPosition) : crossSize};
```
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: You are dealing with a positioned layout here, but the sizes of the items has already been determined by the unpositioned layout algorithm. Don't you need to update that algorithm as well? Maybe I'm missing something though... it's been a long time since I worked on this code.

File: ComponentKit/LayoutComponents/CKStackPositionedLayout.mm (hunk 3 above)
Comment (antigp) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: SpacingBefore always add even in the first item of line. 
style.spacing will not be added in new line. I added a flag setting "first" to YES

Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackPositionedLayout.mm]: Ahh, I see you're setting first above. Got it.

**Comment by ocrickard** [comment]
As I mentioned, I believe this is an incomplete implementation:

https://www.w3.org/TR/css-flexbox-1/#multi_line

At the point you're injecting this code in the algorithm the sizes have been determined according to a single-line layout. The standard specifies that each line should be laid out independently according to the flex-box rules.

What we actually need here is a multi-pass sizing/positioning algorithm. In the unpositioned layout phase we need to determine how many lines we are going to break into, and size each of the lines using the normal algorithm. This will probably take two passes (one sizing pass to line break, one to size independently once we know how many elements will fit on each line).

**Comment by ocrickard** [comment]
I think it'd be awesome to approach this from a TDD perspective. Write a whole bunch of snapshot tests where we would hope to get all the edge-cases covered. Compare the output behavior with the behavior of Chrome to confirm the expected behavior, then tweak the algorithm until we get the expected results.

**Comment by ocrickard** [comment]
I'm going to close the PR for now. Please re-issue a PR after addressing feedback.

**Comment by wy19901227** [comment]
@antigp why not to write a new component named CKFlowLayoutComponent?
